fsspec/implementations/gist.py

```diff
@@ -0,0 +1,211 @@
+import requests
+
+import fsspec
+
+from ..spec import AbstractFileSystem
+from ..utils import infer_storage_options
+from .memory import MemoryFile
+
+
+class GistFileSystem(AbstractFileSystem):
+    """
+    Interface to files in a single GitHub Gist.
+
+    Provides read-only access to a gist's files. Gists do not contain
+    subdirectories, so file listing is straightforward.
+
+    Parameters
+    ----------
+    gist_id : str
+        The ID of the gist you want to access (the long hex value from the URL).
+    sha : str (optional)
+        If provided, fetch a particular revision of the gist. If omitted,
+        the latest revision is used.
+    username : str (optional)
+        GitHub username for authentication (required if token is given).
+    token : str (optional)
+        GitHub personal access token (required if username is given).
+    timeout : (float, float) or float, optional
+        Connect and read timeouts for requests (default 60s each).
+    kwargs : dict
+        Passed to AbstractFileSystem base class.
+    """
+
+    protocol = "gist"
+    gist_url = "https://api.github.com/gists/{gist_id}"
+    gist_rev_url = "https://api.github.com/gists/{gist_id}/{sha}"
+
+    def __init__(
+        self, gist_id, sha=None, username=None, token=None, timeout=None, **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.gist_id = gist_id
+        self.sha = sha  # revision of the gist (optional)
+        if (username is None) ^ (token is None):
+            # Both or neither must be set
+            if username or token:
+                raise ValueError("Auth requires both username and token, or neither.")
+        self.username = username
+        self.token = token
+        # Default timeouts to 60s connect/read if none provided
+        self.timeout = timeout if timeout is not None else (60, 60)
+
+        # We use a single-level "directory" cache, because a gist is essentially flat
+        self.dircache[""] = self._fetch_file_list()
```

Comment:
It is unusual to fetch file listings on init, but I think it makes sense in this case.

Reply:
Yeah I think it's required when you want all the files, but having thought about it a bit there'd also be cases where you want one specific file and would thus be able to forego the file listing, I suspect?

For context I was trying to pull in all files from a gist, but there's no reason why you couldn't populate a single file "file system" of a single file (for a gist it isn't a bad idea, they often are a single file).

Reply:
Hm no turns out there is no single file API endpoint so it would have to do the same metadata request for all files it currently does, or else it wouldn't be able to have the size of the file in question (and since the dircache is immutable now, we wouldn't be able to get it).

Reply:
I implemented it regardless, it's still useful to know you'll get a `FileNotFoundError` if the file isn't in the gist